components/datetime/src/neo.rs

```diff
@@ -803,47 +806,148 @@ impl<R: NeoFormatterMarker> NeoFormatter<R> {
 
     /// Formats a datetime, checking that the calendar system is correct.
     ///
-    /// If the datetime is in neither ISO-8601 nor the same calendar system as the formatter,
+    /// If the datetime is not in the same calendar system as the formatter,
     /// an error is returned.
     ///
-    /// For an example, see [`NeoDateFormatter`].
-    pub fn format<T>(
+    /// # Examples
+    ///
+    /// Mismatched calendars will return an error:
+    ///
+    /// ```
+    /// use icu::calendar::Date;
+    /// use icu::datetime::neo::NeoFormatter;
+    /// use icu::datetime::neo_marker::NeoYearMonthDayMarker;
+    /// use icu::datetime::neo_skeleton::NeoSkeletonLength;
+    /// use icu::datetime::MismatchedCalendarError;
+    /// use icu::locid::locale;
+    ///
+    /// let formatter = NeoFormatter::<NeoYearMonthDayMarker>::try_new(
+    ///     &locale!("en-u-ca-hebrew").into(),
+    ///     NeoSkeletonLength::Long,
+    /// )
+    /// .unwrap();
+    ///
+    /// let date = Date::try_new_gregorian_date(2023, 12, 20).unwrap();
+    ///
+    /// assert!(matches!(
+    ///     formatter.strict_format(&date),
+    ///     Err(MismatchedCalendarError { .. })
+    /// ));
+    /// ```
+    ///
+    /// A time cannot be passed into the formatter when a date is expected:
+    ///
+    /// ```compile_fail
+    /// use icu::calendar::Time;
+    /// use icu::datetime::neo::NeoFormatter;
+    /// use icu::datetime::neo_marker::NeoYearMonthDayMarker;
+    /// use icu::datetime::neo_skeleton::NeoSkeletonLength;
+    /// use icu::locid::locale;
+    ///
+    /// let formatter = NeoFormatter::<NeoYearMonthDayMarker>::try_new(
+    ///     &locale!("es-MX").into(),
+    ///     NeoSkeletonLength::Long,
+    /// )
+    /// .unwrap();
+    ///
+    /// // the trait `for<'a> From<&'a icu::icu_calendar::Time>`
+    /// // is not implemented for `NeoDateInputFields`
+    /// formatter.strict_format(&Time::try_new(0, 0, 0, 0).unwrap());
+    /// ```
+    pub fn strict_format<T>(
         &self,
         datetime: &T,
     ) -> Result<FormattedNeoDateTime, crate::MismatchedCalendarError>
     where
-        T: DateTimeInput<Calendar = AnyCalendar>,
+        T: ?Sized,
+        for<'a> &'a T: IsAnyCalendarKind,
```

Comment:
issue/question: it's unclear to me why this HRTB is necessary at all. Why not have the trait implemented on `T`?

Reply:
The traits need to be implemented on `&T` instead of `T` because `From` moves `T`.

I could use some other trait that's not `From` to avoid this.

Reply:
Yeah that's what I'm suggesting elsewhere, but I can't see why IsAnyCalendarKind needs to be on &T because of From

Reply:
Yeah I can probably remove this HRTB; what is your suggestion for the others?

Reply:
Use a trait that isn't From, and for the InCalendar one you have a trait already and just make it take a reference